[nemo-qml-plugin-systemsettings] Use NemoDBus::Interface rather than QDBusInterface. Contributes to JB#55291 OMP#JOLLA-506

chriadam opened the pull request:
QDBusInterface performs a blocking introspection on construction.
Avoid this by using NemoDBus::Interface instead.

chriadam commented:
Still WIP, do not merge.

pvuorela commented:
Avoiding QDBusInterface is good, but why change all the other standard qt api that already works good? There should be a reason on using dbus wrappers but now all the classes are not even using any QDBusInterface.

chriadam commented:
What are you referring to?  I only changed the QDBusInterface to NemoDBus::Interface (and then, in the cases where we also needed to change QDBusConnection to NemoDBus::Connection because NemoDBus::Interface uses that one, I also updated e.g. signal connections from the QDBusConnection to NemoDBus::Connection -- which is just a simple wrapper anyway).  There are zero other changes in the PR.

chriadam commented:
Well, that wasn't entirely true - there was one case where I had converted a QDBusConnection to a NemoDBus::Connection which was unnecessary.  I've now reverted that change.

chriadam commented:
And added suppression of some spurious warnings from udisks2 dbus calls.

chriadam commented:
No longer WIP, this is now ready for review.  Manual testing steps:
        - plug/unplug USB, see charger status change in status bar
        - wait a few minutes, see battery percentage drop in status bar
        - settings -> language -> change language -> device should reboot.
        - location settings: set custom mode, enable/disable gps, backstepping each time.  Should change connman gps technology enablement status.
        - nfc settings, get enabled/set enabled.
        - user settings, enable / disable guest user, add or remove additional user (NOTE: I only checked the guest user..)
        - storage - check memory card is found

chriadam commented:
@aroditelev review appreciated also

chriadam commented on src/usermodel.cpp:
```diff
@@ -463,138 +495,52 @@ void UserModel::setGuestEnabled(bool enabled)
         qCDebug(lcUsersLog) << "Destroying interface to user-managerd";
-        disconnect(m_dBusInterface, 0, this, 0);
-        m_dBusInterface->deleteLater();
+        delete m_dBusInterface; // TODO: will this also disconnect the signals?
```
@adenexter wondering about this one, I guess internally the connection is from the QDBusConnection to the QObject, rather than from the NemoDBus::Interface, so maybe deleting the interface here won't disconnect signals appropriately.  Or am I just confusing myself?

adenexter replied:
It won't. Interface::connectToSignal() is just QDBusConnection::connect() with fewer arguments.

chriadam replied:
Thanks. Manually added disconnects now.

adenexter commented:
> 
> 
> Avoiding QDBusInterface is good, but why change all the other standard qt api that already works good? There should be a reason on using dbus wrappers but now all the classes are not even using any QDBusInterface.

adenexter commented:
> 
> 
> QDBusInterface performs a blocking introspection on construction. Avoid this by using NemoDBus::Interface instead.

If you just want QDBusInterface without the truly diabolical bits that's QDBusAbstractInterface.

chriadam commented:
> > QDBusInterface performs a blocking introspection on construction. Avoid this by using NemoDBus::Interface instead.
> 
> If you just want QDBusInterface without the truly diabolical bits that's QDBusAbstractInterface.

Is there any reason to prefer QDBusAbstractInterface over NemoDBus::Interface in this case?

Tomin1 commented on src/usermodel.cpp:
```diff
@@ -367,10 +394,15 @@ void UserModel::removeGroups(int row, const QStringList &groups)
+    auto response = m_dBusInterface->call(QStringLiteral("removeFromGroups"), (uint)uid, groups);
+    response->onError([this, uid](const QDBusError &error) {
+        emit removeGroupsFailed(m_uidsToRows.value(uid), getErrorType(error));
+        qCWarning(lcUsersLog) << "Adding user to groups failed:" << error;
```
Would you mind fixing this old mistake? It says "Adding" although this is clearly about removing.

chriadam replied:
Done

Tomin1 replied:
Thanks!

Tomin1 reviewed (commented):
This has a few of these blocking calls (which it had also before, it's not a new thing) so I wonder how much work it would be to refactor those to be asynchronous. But even without that this seems like a good improvement and I didn't see mistakes in the code (apart from that already existing typo that I commented about) so I think it looks good.

I would have probably used xmls and qdbusxml2cpp to get QAbstractDBusInterface subclasses to use those in places that extensively use different D-Bus methods and for the rest just made calls directly with QDBusConnection::asyncCall/call. But I think this works too.

chriadam commented:
> This has a few of these blocking calls (which it had also before, it's not a new thing) so I wonder how much work it would be to refactor those to be asynchronous. But even without that this seems like a good improvement and I didn't see mistakes in the code (apart from that already existing typo that I commented about) so I think it looks good.

There are currently four blocking calls that I can see:
DiskUsageWorker::calculateApkdSize()
DeveloperModeSettings::usbModedGetConfig()
LocationSettings::gpsFlightMode() (and also setGpsFlightMode())
LanguageModel::setSystemLocale() (calls reboot)

The last one should stay synchronous.  In LocationSettings, gpsFlightMode()/setGpsFlightMode() specifically has a synchronous vs asynchronous mode selection, presumably asynchronous is used for UI applications and synchronous is used for e.g. MDM.  The first two are currently designed as synchronous property lookups, so potentially could return some cached value and emit change signal when ready, but probably simpler to leave as-is for now.

chriadam commented:
JP didn't find any regressions with the RPMs I attached to the bug.  Should I merge this one, or are there still some changes which should be made to the PR first?

Tomin1 reviewed (approved):
LGTM. 

Ping @aroditelev

aroditelev commented:
@adenexter your opinion about this MR is needed, since you reopened it.

aroditelev commented:
@pvuorela  please look again

rainemak commented:
Rebased
